Fix QASM examples (#1993)
Qiskit/documentation: docs/guides/primitives-rest-api.mdx

  </TabItem>
</Tabs>

<span id="start-sampler-parms"></span>
## Sampler primitive with REST API and parameterized circuits

### 1. Initialize the account

> Because Qiskit Runtime Sampler is a managed service, you first need to initialize your account. You can then select the device you want to use to run your calculations on.

Reviewer: ```suggestion
Because Qiskit Runtime is a managed service, you first need to initialize your account. You can then select the device you want to use to run your calculations on.
```

Find details on how to initialize your account, view available backends, and invalidate tokens in this [topic](./setup-channel#set-up-to-use-ibm-quantum-platform-with-rest-api).

### 2. Define parameters

```python
import requests
import qiskit_ibm_runtime

> # Imports from Qiskit Runtime

Reviewer: This is noisy, and it's not even true. The line above is from Runtime and the last line is not.
```suggestion
```

from qiskit_ibm_runtime import QiskitRuntimeService
from qiskit.transpiler.preset_passmanagers import generate_preset_pass_manager

> 

Reviewer: ```suggestion
```

from qiskit.qasm3 import dumps
from qiskit import QuantumCircuit
from qiskit.circuit import Parameter
from qiskit import transpile

> service = QiskitRuntimeService(channel='ibm_quantum')

Reviewer: Is `channel` necessary? We don't normally set this, do we?

Author: It's not really necessary, but we specify it in several examples.  If you have access to cloud and IQP, it defaults to Cloud, which can be a surprise.  Since it was in the code I was given, I'll probably just leave it.

> backend = service.backend("your selected backend")

Reviewer: ```suggestion
backend = service.backend("<your selected backend>")
```

Reviewer: Should this be least_busy()?

pm = generate_preset_pass_manager(backend=backend, optimization_level=1)

theta = Parameter('theta')
phi = Parameter('phi')
parameter_values = {'theta': 1.57, 'phi': 3.14}   # In case we want to pass a dictionary
```

### 3. Create a quantum circuit and add parameterized gates

```python
qc = QuantumCircuit(2)

# Add parameterized gates
qc.rx(theta, 0)
qc.ry(phi, 1)
qc.cx(0, 1)
qc.measure_all()

# Draw the original circuit
qc.draw('mpl')

# Get an ISA circuit

> isa_circuit= pm.run(qc)

Reviewer: ```suggestion
isa_circuit = pm.run(qc)
```

```

### 4. Generate QASM 3 code

```python
qasm_str = dumps(isa_circuit)
print("Generated QASM 3 code:")
print(qasm_str)
```

### 5. Run the quantum circuit using Sampler V2 API

<Admonition type="note">

>     The jobs below use [Qiskit Runtime V2 primitives](/migration-guides/v2-primitives). Both `SamplerV2` and `EstimatorV2` take one or more primitive unified blocs (PUBs) as the input. Each PUB is a tuple that contains one circuit and the data broadcasted to that circuit, which can be multiple observables and parameters. Each PUB returns a result.

Reviewer: Should this be `/guides/primitives`? Also link to the API docs for Sampler and Estimator? Link to the PUB file?

    "project": "project1",
    "start_session": "False", # set to False if you just need to run a single job.
    "params": {
        # Choose one option: direct parameter transfer or through a dictionary
        #"pubs": [[qasm_str,[1,2],500]], # primitive unified blocs (PUBs) containing one circuit each. c.f. https://docs.quantum.ibm.com/migration-guides/v2-primitives
        "pubs": [[qasm_str,parameter_values,500]], # primitive unified blocs (PUBs) containing one circuit each. c.f. https://docs.quantum.ibm.com/migration-guides/v2-primitives
        "version": 2 # this defines the version of the Qiskit Runtime Primitive to use, c.f. https://docs.quantum.ibm.com/migration-guides/v2-primitives
}}

response = requests.post(url, headers=headers, json=job_input)

if response.status_code == 200:
    job_id = response.json().get('id')
    print("Job created:",response.text)
else:
    print(f"Error: {response.status_code}")
```

```python
print(response.text)
```

### 6. Check job status and get results

Next, pass the `job_id` to the API:

```python
response_status_singlejob= requests.get(url+'/'+job_id, headers=headers)
response_status_singlejob.json().get('state')
```

> Output

Reviewer: ```suggestion

Output
```

```text
{'status': 'Completed'}
```

> Get job results:

Reviewer: ```suggestion
Get job results:

```

```python
response_result= requests.get(url+'/'+job_id+'/results', headers=headers)

res_dict=response_result.json()

# Get results for the first PUB
counts=res_dict['results'][0]['data']['c']['samples']

print(counts[:20])
```

> Output

Reviewer: ```suggestion

Output

```

```

### 3. Create a quantum circuit and add parameterized gates

```python
qc = QuantumCircuit(2)

# Add parameterized gates
qc.rx(theta, 0)
qc.ry(phi, 1)
qc.cx(0, 1)
qc.measure_all()

# Draw the original circuit
qc.draw('mpl')

# Get an ISA circuit
isa_circuit = pm.run(qc)
```

### 4. Generate QASM 3 code

```python
qasm_str = dumps(isa_circuit)
print("Generated QASM 3 code:")
print(qasm_str)
```

### 5. Run the quantum circuit using Sampler V2 API

<Admonition type="note">
    The jobs below use Qiskit Runtime V2 [primitives](/guides/primitives). Both [`SamplerV2`](/api/qiskit-ibm-runtime/qiskit_ibm_runtime.SamplerV2) and [`EstimatorV2`](/api/qiskit-ibm-runtime/qiskit_ibm_runtime.EstimatorV2) take one or more [primitive unified blocs (PUBs)](/guides/primitive-input-output#pub-overview) as the input. Each PUB is a tuple that contains one circuit and the data broadcasted to that circuit, which can be multiple observables and parameters. Each PUB returns a result.
</Admonition>

```python
import requests

url = 'https://api.quantum-computing.ibm.com/runtime/jobs'
headers = {
    'Content-Type': 'application/json',
    '"Authorization": f"Bearer xxxxxxx",

>     'x-qx-client-application': 'qiskit-version-2/0.39.2/'+'your_application' # specifying the application you might be running from. For an actual Integration project, this option is invaluable to know where jobs are coming from. At this time the "qiskit-version-2/0.39.2/" string is a necessary prefix.

Reviewer: Can you please double check with the SMEs `qiskit-version-2/0.39.2` is still accurate - I'm surprised it's set to 0.39.2?

```suggestion
    # You should replace <your_application> with the name of your app. This is useful to know
    # where jobs are coming from. Note that the prefix "qiskit-version-2/0.39.2/" is necessary.
    'x-qx-client-application': 'qiskit-version-2/0.39.2/<your_application>',
```

Reviewer: I still suggest my change to use an f string and put the comment above the line, split over two lines

Author: Yep - I wasn't ignoring that comment; I was just leaving it open because I'm still looking for a good answer from an SME.

>     }

Reviewer: ```suggestion
}

```

job_input = {
    'program_id': 'sampler',
    "backend": backend,
    "hub": "hub1",
    "group": "group1",
    "project": "project1",
    "start_session": "False", # set to False if you just need to run a single job.
    "params": {
        # Choose one option: direct parameter transfer or through a dictionary
        #"pubs": [[qasm_str,[1,2],500]], # primitive unified blocs (PUBs) containing one circuit each. c.f. https://docs.quantum.ibm.com/migration-guides/v2-primitives
        "pubs": [[qasm_str,parameter_values,500]], # primitive unified blocs (PUBs) containing one circuit each. c.f. https://docs.quantum.ibm.com/migration-guides/v2-primitives
        "version": 2 # this defines the version of the Qiskit Runtime Primitive to use, c.f. https://docs.quantum.ibm.com/migration-guides/v2-primitives
}}

response = requests.post(url, headers=headers, json=job_input)

if response.status_code == 200:
    job_id = response.json().get('id')

>     print("Job created:",response.text)

Reviewer: ```suggestion
    print(f"Job created: {response.text}")
```

else:
    print(f"Error: {response.status_code}")
```

```python
print(response.text)
```

### 6. Check job status and get results

Next, pass the `job_id` to the API:

```python

> response_status_singlejob= requests.get(url+'/'+job_id, headers=headers)

Reviewer: ```suggestion
response_status_singlejob = requests.get(f"{url}/{job_id}", headers=headers)
```

response_status_singlejob.json().get('state')
```

Output

```text
{'status': 'Completed'}
```

Get job results:

```python

> response_result= requests.get(url+'/'+job_id+'/results', headers=headers)

Reviewer: ```suggestion
response_result = requests.get(f"{url}/{job_id}/results", headers=headers)
```

### 2. Define parameters

```python
import requests
import qiskit_ibm_runtime
from qiskit_ibm_runtime import QiskitRuntimeService
from qiskit.transpiler.preset_passmanagers import generate_preset_pass_manager
from qiskit.qasm3 import dumps
from qiskit import QuantumCircuit
from qiskit.circuit import Parameter
from qiskit import transpile

service = QiskitRuntimeService(channel='ibm_quantum')
backend = service.backend("<your selected backend>")

pm = generate_preset_pass_manager(backend=backend, optimization_level=1)

theta = Parameter('theta')
phi = Parameter('phi')
parameter_values = {'theta': 1.57, 'phi': 3.14}   # In case we want to pass a dictionary
```

### 3. Create a quantum circuit and add parameterized gates

```python
qc = QuantumCircuit(2)

# Add parameterized gates
qc.rx(theta, 0)
qc.ry(phi, 1)
qc.cx(0, 1)
qc.measure_all()

# Draw the original circuit
qc.draw('mpl')

# Get an ISA circuit
isa_circuit = pm.run(qc)
```

### 4. Generate QASM 3 code

```python
qasm_str = dumps(isa_circuit)
print("Generated QASM 3 code:")
print(qasm_str)
```

### 5. Run the quantum circuit using Sampler V2 API

<Admonition type="note">

>     The jobs below use Qiskit Runtime V2 [primitives](/guides/primitives). Both [`SamplerV2`](/api/qiskit-ibm-runtime/qiskit_ibm_runtime.SamplerV2) and [`EstimatorV2`](/api/qiskit-ibm-runtime/qiskit_ibm_runtime.EstimatorV2) take one or more [primitive unified blocs (PUBs)](/guides/primitive-input-output#pub-overview) as the input. Each PUB is a tuple that contains one circuit and the data broadcasted to that circuit, which can be multiple observables and parameters. Each PUB returns a result.

Reviewer 2: ```suggestion
    The following jobs use Qiskit Runtime V2 [primitives](/guides/primitives). Both [`SamplerV2`](/api/qiskit-ibm-runtime/qiskit_ibm_runtime.SamplerV2) and [`EstimatorV2`](/api/qiskit-ibm-runtime/qiskit_ibm_runtime.EstimatorV2) take one or more [primitive unified blocs (PUBs)](/guides/primitive-input-output#pub-overview) as the input. Each PUB is a tuple that contains one circuit and the data broadcasted to that circuit, which can be multiple observables and parameters. Each PUB returns a result.
```
